File: src/main/scala/Category.scala
object Category extends App {

  //Composition
  val f: Int => Int = _*2
  val g: Int => Int = _+1
  def `◦`[R,S,T](f: R => S, g: S => T): (R => T) = (x:R) => g(f(x))
  assert(`◦`(f,g)(4) == 9)

  //Properties of composition
  //1.associative
  val h: Int => Int = _+2
  assert(`◦`(`◦`(f,g), h)(4) ==  `◦`(f, `◦`(g,h))(4))

  //2.identity
  def id[T](t:T):T = t
  assert(`◦`(id[Int],f)(4) == f(4))
  assert(`◦`(id[Int],f)(4) == `◦`(f,id[Int])(4))

  //(Scala already has a predefined identity function)
  assert(identity("fe") == "fe")

}

File: src/main/java/fjabj/Category.java
package fjabj;

import java.util.function.Function;

public class Category {

  public static void main(String[] args) {

    var x = 4;

    //Composition
    Function<Integer,Integer> f = i -> i*2;
    Function<Integer,Integer> g = i -> i+1;

    assert((f.andThen(g)).apply(x) != (g.andThen(f)).apply(x));

    //Properties of composition
    //1.associative
    Function<Integer,Integer> h = i -> i+2;

    assert(((f.andThen(g)).andThen(h)).apply(x) == (f.andThen(g.andThen(h))).apply(x));

    //2.identity
    Function id = i -> i;
    assert((f.andThen(id)).apply(x) == id.andThen(f).apply(x));

    //(Java already has a predefined identity function)
    assert(Function.<Integer>identity().apply(x) == x);

    Function<Integer,Integer> functionDerivedFromMethod = Category::myMethod;
    assert(functionDerivedFromMethod.apply(x) == x);

  }

  static int myMethod(int x){
    return x;
  }
}
